proposals/0292-package-registry-service.md

```diff
@@ -596,14 +593,65 @@ in descending order of precedence:
 * Any local configuration (`./.swiftpm/config/registries.json`)
 * Any global configuration file (`~/.swiftpm/config/registries.json`)
 
+#### Specifying credentials for a custom registry
+
+Some servers may require a username and password.
+The user can provide credentials when setting a custom registry
+by passing the `--login` and `--password` options.
+
+When credentials are provided,
+the corresponding object in the `registries.json` file
+includes a `login` key with the passed value.
+If the project's `.netrc` file has an existing entry
+for a given machine and login,
+it's updated with the new password;
+otherwise, a new entry is added.
+If no `.netrc` file exists,
+a new one is created and populated with the new entry.
+
+```console
+$ swift package-registry set https://internal.example.com/ \
+    --login jappleseed --password alpine
+
+$ cat .netrc
+machine internal.example.com
+login jappleseed
+password alpine
+
+$ cat .swiftpm/config/registries.json
+
+{
+  "registries": {
+    "[default]": {
+      "url": "https://internal.example.com"
+      "login": "jappleseed"
+    }
+  },
+  "version": 1
+}
+```
+
+If the user passes the `--login` and `--password` options
+to the `set` subcommand along with the `--global` option,
+the user-level `.netrc` file is updated instead.
+When Swift Package Manager connects to a custom registry,
+it first consults the project's `.netrc` file, if one exists.
+If no entry is found for the custom registry,
+Swift Package Manager then consults the user-level `.netrc` file, if one exists.
+
+If the provided credentials are missing or invalid,
+Swift Package Manager commands like
+`swift package resolve` and `swift package update`
+fail with an error.
+
```

Comment:
@Lukasa @mmarston ptal ^^

Reply:
This UX, the ability to pass `--login` and `--password`, is what I was looking for, so this looks good to me. I have a few questions about the behavior.

1. Does this support using different credentials for different registries served from the same host name (at different paths)? I'm inferring that the answer is yes since the `login` field is added to `registries.json`, which means that the `.netrc` could have more than one entry for the same `machine` with different `login` and the Swift Package Manager would select the one with the matching `login`.
2. Is the user allowed to omit `--password` when passing `--login`? (For example, to specify which login to use if the user already has credentials in their user-level `.netrc`).
3. What credentials will the Swift Package Manager use if the registry redirects to a different host? For example, the a private registry may require credentials and then it may redirect the client to a pre-signed S3 URL. From a security perspective the client shouldn't include the credentials when following a redirect to a different host. Will the client attempt to look up the host of the redirect location in a `.netrc` file?

Reply:
> 1. Does this support using different credentials for different registries served from the same host name (at different paths)?

Yes. A different login / password combination can be specified to add a separate entry in the `.netrc` file.

> 2. Is the user allowed to omit `--password` when passing `--login`

Yes.

> 3. What credentials will the Swift Package Manager use if the registry redirects to a different host?

I'll defer to @tomerd about this specific implementation detail, but my expectation as a user and registry operator would be for any provided credentials (like in a pre-signed S3 URL) to be respected. If that request fails with a 401, I think there's a good case for Swift Package Manager to retry with any available `.netrc` credentials.

Reply:
this all makes sense to me

Reply:
This approach seems reasonable, provided that SwiftPM's default .gitignore file ignores the project .netrc file.

Reply:
@Lukasa I just submitted a PR with that change here: https://github.com/apple/swift-package-manager/pull/3511